lib/MusicBrainz/Server/Sitemap/Incremental.pm

```diff
@@ -217,52 +220,123 @@ sub post_replication_sequence {
     my ($self, $c) = @_;
 
     for my $entity_type (@{ $self->dumped_entity_types }) {
-        # The sitemaps.control columns will be NULL on first run, in which case
-        # we just select all updates for writing.
-        my $all_updates = $c->sql->select_list_of_hashes(
-            "SELECT lm.*
-               FROM sitemaps.${entity_type}_lastmod lm
-              WHERE lm.replication_sequence > COALESCE(
-                (SELECT overall_sitemaps_replication_sequence FROM sitemaps.control), 0)",
+        # In most cases, there should be no need to create batches for the
+        # incremental sitemaps. An exception may occur in cases where the
+        # sitemaps have not run for some time and many changes have
+        # accumulated (MBS-12239).
+        #
+        # The batching code is mostly copied from `build_one_entity` in
+        # `Sitemap::Overall`. However, the logic to always put the last batch
+        # in its own sitemap is not performed here, since we're not as
+        # concerned about which sitemap a URL starts in for an *incremental*
+        # dump. (As mentioned previously, it's very unlikely there will be
+        # more than one batch, anyway, due to the code that attempts to merge
+        # adjacent ones.)
+
+        # Find the counts in each potential batch of 50,000.
+        my $raw_batches = $c->sql->select_list_of_hashes(
+            <<~"SQL",
+                SELECT batch, count(id) FROM (
+                    SELECT lm.id, ceil(lm.id / ?::float) AS batch
+                    FROM sitemaps.${entity_type}_lastmod lm
+                    WHERE lm.replication_sequence > COALESCE(
+                        (SELECT overall_sitemaps_replication_sequence
+                        FROM sitemaps.control),
+                        0
+                    )
+                ) q
+                GROUP BY batch
+                ORDER BY batch ASC
+                SQL
+            $MAX_SITEMAP_SIZE,
         );
 
-        my %updates_by_suffix_key = partition_by {
-            $_->{sitemap_suffix_key}
-        } @{$all_updates};
+        my $batch = {count => 0, batches => []};
+        my @batches = ($batch);
+        for my $raw_batch (@$raw_batches) {
+            # Add this potential batch to the previous one if the sum will
+            # come out less than 50,000. Otherwise create a new batch and
+            # push the previous one onto the list.
```

Comment:
```suggestion
            # Add this potential batch to the previous one if the sum will
            # come out less than $MAX_SITEMAP_SIZE. Otherwise create a new
            # batch and push the previous one onto the list.
```